### PR TITLE
feat: register & deregister region failure detectors actively

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4161 

## What's changed and what's your intention?

- Register failure detector during create table procedure. The datanode may crash without sending a heartbeat that contains information about newly created regions, which may prevent the RegionSupervisor from detecting failures in these newly created regions.
- Register failure detector during rollback region migration procedure. The original failure detector was removed once the procedure was triggered. Therefore, we need to register the failure detector for the failed region during rollback.
- Deregister failure detectors during drop table procedure. Once the regions were dropped, subsequent heartbeats no longer include these regions. Therefore, we need to remove the failure detectors for these dropped regions.
    

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
	- Introduced region failure detection for enhanced table creation and region management processes.

- **Improvements**
	- Updated handling of region failure detectors across various components.
	- Added initialization and registration of failure detectors in multiple workflows.

- **Configuration**
	- Adjusted `acceptable_heartbeat_pause` value from `"3000ms"` to `"10000ms"` for better failure detection timing.

- **Tests**
	- Updated test setups to include failure detector initialization and validated event handling.
	- Modified random generation parameters in failover tests for optimized data handling.

### CODE REVIEW

**Comment by WenyXu** [comment]
@coderabbitai resolve

File: src/common/meta/src/ddl.rs [hunk 1]
```diff
@@ -102,6 +102,28 @@ pub struct TableMetadata {
     pub region_wal_options: HashMap<RegionNumber, String>,
 }
 
+pub type RegionFailureDetectorControllerRef = Arc<dyn RegionFailureDetectorController>;
+
+#[async_trait::async_trait]
+pub trait RegionFailureDetectorController: Send + Sync {
+    /// Registers failure detectors for the given identifiers.
+    async fn register_failure_detectors(&self, ident: Vec<(ClusterId, DatanodeId, RegionId)>);
```
Comment (fengjiachun) [inline src/common/meta/src/ddl.rs:110]: How about:
```
type RegionIdent = (ClusterId, DatanodeId, RegionId)
```

Comment (fengjiachun) [inline src/common/meta/src/ddl.rs:110]:
```suggestion
    async fn register_failure_detectors(&self, idents: Vec<(ClusterId, DatanodeId, RegionId)>);
```

File: src/common/meta/src/ddl.rs [hunk 2]
```diff
@@ -102,6 +102,28 @@ pub struct TableMetadata {
...
+
+    /// Deregisters failure detectors for the given identifiers.
+    async fn deregister_failure_detectors(&self, ident: Vec<(ClusterId, DatanodeId, RegionId)>);
```
Comment (fengjiachun) [inline src/common/meta/src/ddl.rs:113]:
```suggestion
    async fn deregister_failure_detectors(&self, idents: Vec<(ClusterId, DatanodeId, RegionId)>);
```

File: src/common/meta/src/ddl.rs [hunk 3]
```diff
@@ -102,6 +102,28 @@ pub struct TableMetadata {
     pub region_wal_options: HashMap<RegionNumber, String>,
 }
 
+pub type RegionFailureDetectorControllerRef = Arc<dyn RegionFailureDetectorController>;
+
+#[async_trait::async_trait]
+pub trait RegionFailureDetectorController: Send + Sync {
```
Comment (fengjiachun) [inline src/common/meta/src/ddl.rs:108]: Detailed comments are needed.

**Comment by WenyXu** [comment]
@fengjiachun PTAL

File: src/meta-srv/src/procedure/region_migration.rs [hunk 4]
```diff
@@ -183,6 +187,7 @@ impl ContextFactory for DefaultContextFactory {
             volatile_ctx: self.volatile_ctx,
             table_metadata_manager: self.table_metadata_manager,
             opening_region_keeper: self.opening_region_keeper,
+            region_failure_detector_controller: self.region_failure_detector_controller.clone(),
```
Comment (MichaelScofield) [inline src/meta-srv/src/procedure/region_migration.rs:190]: Seems no need to clone here.

File: src/meta-srv/src/region/supervisor.rs [hunk 5]
```diff
@@ -259,6 +304,21 @@ impl RegionSupervisor {
         info!("RegionSupervisor is stopped!");
     }
 
+    async fn register_failure_detectors(&self, idents: Vec<Ident>) {
+        let ts_millis = current_time_millis();
+        for ident in idents {
+            // The corresponding region has `acceptable_heartbeat_pause_millis` to send heartbeat from datanode.
+            self.failure_detector
+                .maybe_init_region_failure_detector(ident, ts_millis);
```
Comment (MichaelScofield) [inline src/meta-srv/src/region/supervisor.rs:312]: `maybe_init_region_failure_detector` will insert a heartbeat, but `register_failure_detectors` could be called outside of heartbeat processing path. So there might be a "fake" heartbeat. What's the consequence of doing that?

File: src/common/meta/src/ddl.rs [hunk 6]
```diff
@@ -102,6 +102,31 @@ pub struct TableMetadata {
     pub region_wal_options: HashMap<RegionNumber, String>,
 }
 
+pub type RegionFailureDetectorControllerRef = Arc<dyn RegionFailureDetectorController>;
+
+/// Used for actively registering Region failure detectors.
+///
+/// Ensuring the Region Supervisor can detect Region failures without relying on the first heartbeat from the datanode.
+#[async_trait::async_trait]
+pub trait RegionFailureDetectorController: Send + Sync {
```
Comment (fengjiachun) [inline src/common/meta/src/ddl.rs:111]: @coderabbitai

**Comment by fengjiachun** [comment]
@coderabbitai Hi

File: src/common/meta/src/ddl.rs (hunk 6 above)
Comment (fengjiachun) [inline src/common/meta/src/ddl.rs:111]: @coderabbitai Is there a better name to replace the `RegionFailureDetectorController`

File: src/meta-srv/src/region/supervisor.rs (hunk 5 above)
Comment (WenyXu) [reply src/meta-srv/src/region/supervisor.rs:312]: > What's the consequence of doing that?

If the datanode didn't send the heartbeat promptly, the region failover will be trigger again.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/meta-srv/src/region/failure_detector.rs [hunk 7]
```diff
@@ -51,7 +51,7 @@ impl RegionFailureDetector {
         }
     }
 
-    /// Returns [PhiAccrualFailureDetector] of the specific ([DatanodeId],[RegionId]).
+    /// Returns [`PhiAccrualFailureDetector`] of the specific [`Ident`].
```
Comment (killme2008) [inline src/meta-srv/src/region/failure_detector.rs:54]:
```suggestion
    /// Returns `[PhiAccrualFailureDetector]` of the specific `[Ident]`.
```

File: src/meta-srv/src/region/failure_detector.rs [hunk 8]
```diff
@@ -61,6 +61,21 @@ impl RegionFailureDetector {
             .or_insert_with(|| PhiAccrualFailureDetector::from_options(self.options))
     }
 
+    /// Returns A mutable reference to the [`PhiAccrualFailureDetector`] for the specified [`Ident`].
```
Comment (killme2008) [inline src/meta-srv/src/region/failure_detector.rs:64]: Correct the comment, I don't want to repeat it.

File: src/meta-srv/src/region/failure_detector.rs [hunk 9]
```diff
@@ -51,7 +51,7 @@ impl RegionFailureDetector {
         }
     }
 
-    /// Returns [PhiAccrualFailureDetector] of the specific ([DatanodeId],[RegionId]).
+    /// Returns [`PhiAccrualFailureDetector`] of the specific [`Ident`].
     pub(crate) fn region_failure_detector(
         &self,
         ident: Ident,
```
Comment (killme2008) [inline src/meta-srv/src/region/failure_detector.rs:57]: `Ident` is really a terrible name, what about changing it to `DetectingRegion`?

File: src/meta-srv/src/region/supervisor.rs [hunk 10]
```diff
@@ -178,12 +203,53 @@ pub struct RegionSupervisor {
...
+impl RegionFailureDetectorControl {
+    pub(crate) fn new(sender: Sender<Event>) -> Self {
+        Self { sender }
+    }
+}
+
+#[async_trait::async_trait]
+impl RegionFailureDetectorController for RegionFailureDetectorControl {
+    async fn register_failure_detectors(&self, ident: Vec<Ident>) {
+        if self
```
Comment (killme2008) [inline src/meta-srv/src/region/supervisor.rs:221]: Better to print the error.
```rust
if let Err(e) = .... {
    error!(e; "RegionSupervisor is stop receiving heartbeat");
}

File: src/meta-srv/src/region/supervisor.rs [hunk 11]
```diff
@@ -178,12 +203,53 @@ pub struct RegionSupervisor {
...
+    }
+
+    async fn deregister_failure_detectors(&self, ident: Vec<Ident>) {
+        if self
+            .sender
+            .send(Event::DeregisterFailureDetectors(ident))
+            .await
+            .is_err()
+        {
+            error!("RegionSupervisor is stop receiving heartbeat");
```
Comment (killme2008) [inline src/meta-srv/src/region/supervisor.rs:238]: The same suggestion as above.

File: src/meta-srv/src/region/supervisor.rs [hunk 12]
```diff
@@ -420,21 +483,25 @@ pub(crate) mod tests {
...
 
-        RegionSupervisor::new(
-            Default::default(),
-            Duration::from_secs(1),
-            selector_context,
-            selector,
-            region_migration_manager,
-            kv_backend,
-            peer_lookup,
+        let (tx, rx) = tokio::sync::mpsc::channel(1024);
```
Comment (killme2008) [inline src/meta-srv/src/region/supervisor.rs:486]: The magic number 1024 appears many times in this PR. I am not sure it's valuable to be configured or made into a constant if we want to change them in the future.

**Review by killme2008 — COMMENTED** [review]
Another suggestion: can we add some tests to cover these cases?

File: src/common/meta/src/ddl/create_table.rs [hunk 13]
```diff
@@ -265,16 +265,38 @@ impl CreateTableProcedure {
...
+            .collect::<Vec<_>>();
+        let table_route = TableRouteValue::Physical(physical_table_route);
         manager
             .create_table_metadata(raw_table_info, table_route, region_wal_options)
             .await?;
+        // Notifies RegionSupervisor to register failure detector of new created regions.
+        //
+        // The datanode may crash without sending a heartbeat that contains information about newly created regions,
+        // which may prevent the RegionSupervisor from detecting failures in these newly created regions.
+        self.context
```
Comment (killme2008) [inline src/common/meta/src/ddl/create_table.rs:296]: I'd like to wrap these logic into a function like `register_region_failure_detectors`, and make the `on_create_metadata` clean and smaller.

**Comment by WenyXu** [comment]
> Another suggestion: can we add some tests to cover these cases?

Yes, tests will be introduced in following PRs

File: src/meta-srv/src/procedure/region_migration.rs [hunk 14]
```diff
@@ -154,22 +155,25 @@ pub struct DefaultContextFactory {
...
     table_metadata_manager: TableMetadataManagerRef,
     opening_region_keeper: MemoryRegionKeeperRef,
+    region_failure_detector_controller: RegionFailureDetectorControllerRef,
     mailbox: MailboxRef,
     server_addr: String,
 }
 
 impl DefaultContextFactory {
-    /// Returns an [ContextFactoryImpl].
+    /// Returns an [`DefaultContextFactory`].
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/region_migration.rs:164]: Should always be [`somthing`], check:
https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#link-all-the-things

[the remark this answers is not in the document]
Comment (killme2008) [reply src/meta-srv/src/procedure/region_migration.rs:164]: Hmm.. I still can't remember it.

**Comment by WenyXu** [comment]
@killme2008 @MichaelScofield PTAL

File: src/meta-srv/src/region/supervisor.rs [hunk 15]
```diff
@@ -526,4 +595,31 @@ pub(crate) mod tests {
             }
         }
     }
+
+    #[tokio::test]
+    async fn test_region_failure_detector_controller() {
+        let (mut supervisor, sender) = new_test_supervisor();
+        let controller = RegionFailureDetectorControl::new(sender.clone());
+        tokio::spawn(async move { supervisor.run().await });
+        let ident = (0, 1, RegionId::new(1, 1));
```
Comment (killme2008) [inline src/meta-srv/src/region/supervisor.rs:604]: still, some names like `ident`, maybe change them to `region`?

Comment (WenyXu) [reply src/meta-srv/src/region/supervisor.rs]: Renamed in [40b24d5](https://github.com/GreptimeTeam/greptimedb/pull/4223/commits/40b24d501283203c23704a833d1a21c142533a75)

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**